### PR TITLE
feat: Auth scheme changes

### PR DESCRIPTION
**Pull request by sichanyoo**
## Issue \#

Issue: https://github.com/awslabs/aws-sdk-swift/issues/1144

Corresponding PR in `aws-sdk-swift`: https://github.com/awslabs/aws-sdk-swift/pull/1174

## Description of changes

Adds protocols and structs for auth scheme abstraction.
- `IdentityResolverConfiguration` protocol
  - This is the protocol for the object that's passed to an auth scheme's `identityResolver` method. In essence, this object is a smaller "sub-config" object that contains identity resolvers configured on service client config object. It takes in a type of identity, then returns an identity resolver (if any is configured) that returns the indicated type of identity.
- `DefaultIdentityResolverConfiguration` struct
- `AuthOption` struct
- `AuthScheme` protocol 
- `AuthSchemeResolver` protocol
- `AuthSchemeResolverParameters` protocol
- `SelectedAuthScheme` struct

Adds new setters and getters to `HttpContext` to include applicable auth scheme things needed for all operation middleware that use middleware context.

## Scope

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: Sources/ClientRuntime/Identity/IdentityAPI/Identity.swift [hunk 1]
```diff
@@ -11,3 +11,13 @@ import Foundation
...
     var expiration: Date? { get }
 }
+
+// Identity v. IdentityT v. IdentityType
+// - Identity is the protocol that all identity types must conform to.
+// - IdentityT is the associated type / generic type name used by protocols like IdentityResolver and Signer.
+// - IdentityType is the enum that's used by IdentityResolverConfiguration to return correct type of identity resolver
+//   for the given auth scheme. E.g., SigV4AuthScheme has idType field as .aws. And identityResolver method in SigV4AuthScheme
+//   returns an identity resolver that returns identity of type .aws. 
+public enum IdentityType: CaseIterable {
```
Comment (jbelkins) [inline Sources/ClientRuntime/Identity/IdentityAPI/Identity.swift:21]: Is the `IdentityType` name used in SRA?  If not, I'd perfer to find a type name that doesn't include the word "Type"

**Review by jbelkins — CHANGES_REQUESTED** [review]

File: Sources/ClientRuntime/Auth/HTTPAuthAPI/AuthOption.swift [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+// Copyright Amazon.com Inc. or its affiliates.
+// All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import Foundation
+
+public struct AuthOption {
+    let schemeId: String
```
[the remark this answers is not in the document]
Comment (sichanyoo) [reply Sources/ClientRuntime/Auth/HTTPAuthAPI/AuthOption.swift:11]: Changed as suggested

File: Sources/ClientRuntime/Auth/HTTPAuthAPI/AuthSchemeResolver.swift [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+// Copyright Amazon.com Inc. or its affiliates.
+// All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import Foundation
+
+public protocol AuthSchemeResolver {
+    func resolveAuthScheme(params: AuthSchemeResolverParameters) -> Array<AuthOption>
```
[the remark this answers is not in the document]
Comment (sichanyoo) [reply Sources/ClientRuntime/Auth/HTTPAuthAPI/AuthSchemeResolver.swift:11]: Changed as suggested

File: Sources/ClientRuntime/Auth/HTTPAuthAPI/AuthSchemeResolver.swift [hunk 4]
```diff
@@ -0,0 +1,13 @@
+//
+// Copyright Amazon.com Inc. or its affiliates.
+// All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import Foundation
```
[the remark this answers is not in the document]
Comment (sichanyoo) [reply Sources/ClientRuntime/Auth/HTTPAuthAPI/AuthSchemeResolver.swift:8]: Fixed as pointed out

File: Sources/ClientRuntime/Auth/HTTPAuth/DefaultIdentityResolverConfiguration.swift [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+public struct DefaultIdentityResolverConfiguration: IdentityResolverConfiguration {
+    let credentialsProvider: (any IdentityResolver)?
+
+    public init(configuredIdResolvers: Attributes) {
+        self.credentialsProvider = configuredIdResolvers.get(key: AttributeKeys.awsIdResolver) ?? nil
+    }
+
+    func getIdentityResolver(identityType: IdentityType) -> (any IdentityResolver)? {
+        switch identityType {
+        case .aws:
```
[the remark this answers is not in the document]
Comment (sichanyoo) [reply Sources/ClientRuntime/Auth/HTTPAuth/DefaultIdentityResolverConfiguration.swift:19]: There can be bearer token, api key, etc. added here as we support more auth schemes.

File: Sources/ClientRuntime/Identity/IdentityAPI/Identity.swift (hunk 1 above)
Comment (sichanyoo) [reply Sources/ClientRuntime/Identity/IdentityAPI/Identity.swift]: Changed to `IdentityKind` to avoid using the word "Type"

Comment (sichanyoo) [reply Sources/ClientRuntime/Identity/IdentityAPI/Identity.swift]: Also, it's not a name used in SRA - SRA, based on Java, uses `IdentityObject::class.java` to determine the type.

**Review by dayaffe — APPROVED** [review]
Looks good

File: Sources/ClientRuntime/Auth/HTTPAuthAPI/IdentityResolverConfiguration.swift [hunk 6]
```diff
@@ -0,0 +1,10 @@
+//
+// Copyright Amazon.com Inc. or its affiliates.
+// All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+protocol IdentityResolverConfiguration {
+    func getIdentityResolver(identityKind: IdentityKind) -> (any IdentityResolver)?
```
[the remark this answers is not in the document]
Comment (sichanyoo) [reply Sources/ClientRuntime/Auth/HTTPAuthAPI/IdentityResolverConfiguration.swift:9]: This function is to always return a single concrete type that conforms to `IdentityResolver`, but it's dependent on which kind of identity is passed in as the argument. Also, having `some` keyword in protocols is not allowed in Swift. With these two factors combined, 'any' is used here.

**Review by jbelkins — APPROVED** [review]

**Pull request merged**